CI: Test only changed modules and those that depend on them

stefanv opened the pull request:
Credit for this feature goes to @mdhaber.

The purpose of the change is to vastly accelerate CI.
The method is to detect which modules are touched by a PR, and to only test them and their dependencies.

The dependency graph is calculated by (in separate subprocesses) importing each module in skimage, and seeing which modules get brought in by it. These are considered its dependencies.
From the graph, we can answer: if I touch module M, which modules depend on it?

E.g., if I touch the color submodule, I also need to test `skimage.io` and `skimage.segmentation`, since `skimage.color` is imported by those submodules. If I touch `skimage._shared`, I need to test *all* submodules.

There may be exceptions and special cases which are not currently taken into consideration.

This PR aims to measure the speed-up that can be achieved, consider the robustness of the approach, and identify any oversights.

----

**Success!**: [CI run](https://github.com/scikit-image/scikit-image/actions/runs/13808828504/job/38625504451?pr=7749), total runtime 3m49s, of which 2m12s build, 30s for dependency detection and testing

mdhaber committed:
ENH: import graph generator

stefanv committed:
Refactor Matt's tool into stand-alone module

stefanv committed:
Add --detect-dependency flag to spin

stefanv committed:
Add PR CI job: detect changed modules, test their dependencies

stefanv committed:
Fix linting error

stefanv committed:
Test autodepends using spin

stefanv committed:
Fetch main and PR branches, fix pytest config path

stefanv committed:
Make editable install before testing

stefanv committed:
Fix typo

stefanv committed:
Combine build and test deps

stefanv committed:
Run tests separately so we can time

lucascolley commented on tools/module_dependencies.py:
See https://github.com/scipy/scipy/pull/22671 for an alternative to this script

lucascolley replied:
Yeah, not blocking!

mdhaber commented:
Thanks @stefanv! This looks much cleaner and more general than what I sent : ) 
As we discussed, an important exception is when functions perform an import at `runtime`. I think it's ok to add a mechanism for that later, since:
- It's relatively rare 
- We will catch these situations when partial tests pass but full tests fail
- It's only a problem to skip tests that would fail. 

That said, it's possible that tach (gh-22671) considers of this sort of dependency already? So we might be able to fix it automatically. 

It looks like you committed the NPZ file (which maybe would be better as plain text to avoid unreadable diffs?). I imagine that was intentional because it can be used to check whether new dependencies have been added. I didn't see it, but does this already perform that check?

stefanv commented:
> It looks like you committed the NPZ file (which maybe would be better as plain text to avoid unreadable diffs?).

That file came from you, actually ;) 1f758cd67f7a19e78ded4f57381a5c8e856bd3e8

stefanv committed:
Remove npz artifact

stefanv commented:
That said, I now understand what you are hinting at: that it would be useful to also use this to detect dependency changes. The calculation is so quick, currently, that we can just do a checkout, save the dependency matrix, and then compare it with the matrix on the PR branch.

lucascolley commented:
> Maybe it's indeed just a yaml file: submodule -> dependencies.

*or a TOML file (wink wink https://github.com/scipy/scipy/pull/22671 :) )*

stefanv commented:
> _or a TOML file (wink wink [scipy/scipy#22671](https://github.com/scipy/scipy/pull/22671) :) )_

:) Yeah, I saw that. I am not super enthused about TOML, until 1.1 comes out which gives us proper multi-line dictionaries. But whatever works.

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,82 @@
+    Each row represents the dependencies of one subpackage.
+
+    """
+    mods = sorted(_pkg_modules())  # sort for stable matrix
```
This is sorted in three places but used as a set only in `_import_dependencies`. Might be simpler to do the sorting in `_pkg_modules` and convert back into a set if needed only in `_import_dependencies`. I don't think this will be noticeable in performance either way; this is just for simplicity.

mdhaber replied:
Yeah, the tradeoff in performance vs simplicity. Of course, it could return both as a compromise. I guess I value the simplicity (to get something adopted) more than the performance.

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,82 @@
+    importlib.import_module(module)
+    importlib.import_module(f"{module}.tests")
+
+    pkg_modules = _pkg_modules()
+    pkg_sys_modules = {mod for mod in sys.modules if package in mod}
+
+    return pkg_modules & pkg_sys_modules
```
Assuming `_pkg_modules()` ends up returning a sorted list, can this just be:
```suggestion
    return set(_pkg_modules()) & set(sys.module)
```
?

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,82 @@
+    """Return the set of modules that depend on any of the given modules."""
+    changed_modules = modules
+
+    A = dependency_graph().astype(bool)
```
Maybe make it boolean from the start. I only had it as type `int` because it was more compact to print during testing.

stefanv committed:
Add TOML export of dependencies

Internal refactoring:

- Single function to provide coherent modules and modules index list
- Sort modules list internally, always
- Correct type annotations

stefanv committed:
Fix spin dependency processing

stefanv committed:
Avoid cross-contaminating sys.modules by using one worker per submodule

stefanv committed:
Correctly grab test submodules

stefanv committed:
Roll back test change to color module

stefanv committed:
Modify morphology module as a test

stefanv committed:
Get rid of demonic spawning

stefanv committed:
Try without installing

stefanv committed:
Set PYTHONPATH from spin, so it can find spin-built skimage

stefanv commented:
OK, we caught a real bug trigger by `spin test skimage/util/tests/test_lookfor.py`

stefanv commented:
Caused by running `lookfor` when `SimpleITK` is installed :thinking:

stefanv committed:
Remove SimpleITK to pass lookfor test

stefanv committed:
Merge remote-tracking branch 'origin/main' into test-changed-module-dependencies

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,149 @@
+            f"Yikes! No package modules present at this point, but seeing {bad_sys_modules}!"
+        )
+
+    importlib.import_module(module)
```
Erg... Test command is like:

```$ /Library/Frameworks/Python.framework/Versions/3.13/bin/python -P -m pytest --config-file /Users/runner/work/scikit-image/scikit-image/pyproject.toml --pyargs skimage.feature skimage.segmentation skimage.util```

I would have expected it to include `skimage.morphology` : /

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,149 @@
+            pass
+
+    pkg_modules, _ = _pkg_modules()
+    pkg_sys_modules = {mod for mod in sys.modules if mod.startswith(f"{package}.")}
+
+    return set(pkg_modules) & pkg_sys_modules
```
I think the same comment is left over. Can this just be:
```suggestion
    return set(pkg_modules) & set(sys.modules)
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,149 @@
+    for mod in test_modules:
+        try:
+            importlib.import_module(mod)
+        except (ImportError, pytest.skip.Exception):
```
What triggers this?

mdhaber replied:
I see. Worth a comment.
```suggestion
        except (ImportError, pytest.skip.Exception):
        # `pytest.importorskip` raises if library isn't installed
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,149 @@
+    # This is a workaround for identifying the test modules
+    # associated with an skimage submodule.
+    #
+    # Other libraries would need to do something different, and
+    # we should think how to generalize.
```
~~Is this because you don't have any sub-sub-packages that have their own tests?~~ It looks like you do. WIll this catch `filters.rank.tests`?

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -18,30 +21,66 @@
+    for mod in pkg_mods:
+        del sys.modules[mod]
 
+    return submodules, submodule_idx
```
```suggestion
    return submodules
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -18,30 +21,66 @@
+    -------
+    submodules : list
+        Sorted list of package submodules.
+    submodule_idx : dict
+        Mapping of submodule to integer, its index
+        in `submodules`.
```
```suggestion
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -18,30 +21,66 @@
 
 @functools.cache
-def _pkg_modules() -> list[str]:
+def _pkg_modules() -> tuple[list[str], dict[str, int]]:
```
```suggestion
def _pkg_modules() -> tuple[list[str]:
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -18,30 +21,66 @@
+        except (ImportError, pytest.skip.Exception):
+            pass
+
+    pkg_modules, _ = _pkg_modules()
```
```suggestion
    pkg_modules = _pkg_modules()
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -50,33 +89,61 @@ def dependency_graph():
     """
-    mods = sorted(_pkg_modules())  # sort for stable matrix
-    mods_idx = _pkg_modules_index()
+    mods, mods_idx = _pkg_modules()
```
```suggestion
    mods = _pkg_modules()
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -50,33 +89,61 @@ def dependency_graph():
         for k, dependencies in enumerate(mod_deps):
             for mod in dependencies:
-                A[k, mods_idx[mod]] = 1
+                A[k, mods_idx[mod]] = True
```
```suggestion
                A[k, mods.index(mod)] = True
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -50,33 +89,61 @@ def dependency_graph():
 
-def modules_dependent_on(modules: set[str] | list[str]) -> set[str]:
+def dependency_toml():
+    mods, mods_idx = _pkg_modules()
```
```suggestion
    mods = _pkg_modules()
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -50,33 +89,61 @@ def dependency_graph():
     all_dependent_mods = []
     for changed_mod in changed_modules:
-        dependent_mods = pkg_mods[A[:, pkg_mods_idx[changed_mod]]]
+        dependent_mods = pkg_mods_arr[A[:, pkg_mods_idx[changed_mod]]]
```
```suggestion
        dependent_mods = pkg_mods_arr[A[:, pkg_mods.index(changed_mod)]]
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -50,33 +89,61 @@ def dependency_graph():
 
-    pkg_mods = np.array(sorted(_pkg_modules_index()), dtype=object)
-    pkg_mods_idx = _pkg_modules_index()
+    pkg_mods, pkg_mods_idx = _pkg_modules()
```
```suggestion
    pkg_mods = _pkg_modules()
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,149 @@
+    # Sort entries, so that all adjacency matrix calculations are stable
+    submodule_idx = {mod: index for index, mod in enumerate(submodules)}
+
+    # # Clear out sys.modules for when we spawn our import detector
```
Did you find out that there was no way to spawn completely independent processes, or should I look into it?

stefanv replied:
I think the choice of how exactly they spawn is platform-dependent, so not sure how much control we have using `subprocessing`. We could use a different mechanism, but probably not necessary unless we discover a problem with the current approach.

stefanv commented on tools/module_dependencies.py:
```diff
@@ -18,30 +21,66 @@
-    return (members | set(included_mods_full)) - set(excluded_mods_full)
+    submodules = sorted((members | included_mods_full) - excluded_mods_full)
 
+    # Sort entries, so that all adjacency matrix calculations are stable
+    submodule_idx = {mod: index for index, mod in enumerate(submodules)}
```
Speed really isn't an issue with these small data structures, so I'm all for simplicity.

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,149 @@
+    n = len(mods)
+    A = np.zeros((n, n), dtype=bool)
+
+    with multiprocessing.Pool(maxtasksperchild=1) as p:
```
Just to check my understanding of what the problems were and how you fixed them:
- The original script would do e.g. `import skimage.color.tests`, but this didn't actually import `test_` modules. The consequence was failing to run some tests that should have been run.
  - You fixed this by looking for all the `test_` modules and importing them individually.
  - Another possibility we discussed was adding the relevant modules to the test folder's `__init__.py`, but this would require some manual setup, and could get out of sync if modules are added or removed.
  - A third possibility harvesting the information from `meson.build` files within the test folders - or I think you suggested having `meson` create a list for us. But I think the current approach is fine.
- With `multiprocess.Pool`, a given worker might be reused for multiple tasks, so `sys.modules` was sometimes contaminated with the results of an import from a previous task. This resulted in running different sets of tests on different runs.
  - It looks like you fixed that with `maxtasksperchild=1` here?
  - For future reference, was this happening with the original script? That is, was `subprocess` starting a completely fresh process for each module, or did it also re-use processes? (looks like you've partially answered that [here](https://github.com/scikit-image/scikit-image/pull/7749#discussion_r1994079948))

stefanv replied:
Yes, all correct, Matt. ``maxtasksperchild=1` did fix the contamination, the result of running multiple import tasks on a single worker.

The clearing I know needs to happen, because I put an assert in to ensure that none of skimage is in `sys.modules` when we start our scan; and that raises when we don't clear `sys.modules` first.

I think the original script used `subprocess.run`, which I think is safe from the problem we have when using `multiprocessing`. Multiprocessing relies on forking the process, which I believe copies the memory state.

There are some problems with that original approach, though: it requires a standalone script to run, and there's no easy way to read the results back into Python, without parsing stdout.

mdhaber replied:
`multiprocessing.Pool` is clearly a better way; just making sure I'm understanding. Thanks!

stefanv replied:
Well, it does have that sys.modules caveat; so as long as that's not causing us trouble. It is easier to get results from it, that's true.

mdhaber committed:
MAINT: make _pkg_modules return only list

mdhaber committed:
MAINT: suggest adjustment to modules_dependent_on

stefanv committed:
Spawn, instead of forking, new workers

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,146 @@
+    n = len(mods)
+    A = np.zeros((n, n), dtype=bool)
+
+    multiprocessing.set_start_method('spawn')
```
```suggestion
    multiprocessing.set_start_method('spawn', force=True)
```

stefanv committed:
imap implies chunksize=1

mdhaber committed:
Merge remote-tracking branch 'stefan/test-changed-module-dependencies' into gh7749

stefanv committed:
Merge pull request #20 from mdhaber/gh7749

MAINT: make _pkg_modules return only list

mdhaber commented:
Note to self: get sub-sub package tests imported.

mdhaber committed:
MAINT: make test imports recursive

stefanv committed:
Merge pull request #21 from mdhaber/gh7749b

MAINT: make test imports recursive

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,127 @@
+def dependency_graph():
+    """Calculate the dependency graph of the current module.
+
+    Each row represents the dependencies of one subpackage.
```
I think this is correct, as long as it is not ambiguous. To avoid any ambiguity:
```suggestion
    Each row represents the dependencies of one subpackage.
    `A[i, j]` is True if module `i` depends on (imports) module `j`.
```
`j` is a dependency of `i`.
`i` is dependent on `j`.

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,127 @@
+
+    with multiprocessing.Pool(maxtasksperchild=1) as p:
+        mod_deps = p.imap(_import_dependencies, mods)
+        for k, dependencies in enumerate(mod_deps):
+            for mod in dependencies:
+                A[k, mods.index(mod)] = True
```
```suggestion
        for i, dependencies in enumerate(mod_deps):
            for mod in dependencies:
                j = mods.index(mod)
                A[i, j] = True
```
Just to align this with my comment about the docstring.

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,127 @@
+    A = dependency_graph()
+
+    toml = []
+    for i, mod in enumerate(mods):
+        dependency_names = mods_arr[A[:, i]].tolist()
```
Let's consistently use `j` as the subscript for columns (unless you prefer `k` for rows and `i` for columns, which is what is currently here)
```suggestion
    for j, mod in enumerate(mods):
        dependency_names = mods_arr[A[:, j]].tolist()
```

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,127 @@
+    for i, mod in enumerate(mods):
+        dependency_names = mods_arr[A[:, i]].tolist()
+
+        toml.append({"modules": {"path": mod, "depends_on": dependency_names}})
```
I think the English interpretation is reversed here. 
- Each row `i` indicates all the modules that module `i` depends on.
- Each column `j` indicates all the modules that depend on module `j`.
- We are extracting columns here, so the list of modules we get all depend on module `j`.
- So if "path" is module `j`, it is _depended on by_ all these modules.

```suggestion
        toml.append({"modules": {"path": mod, "depended_on_by": dependency_names}})
```
In other words, if we change module `j`, we need to run tests for the modules given by `dependency_names`. Right?  Alternatively:

```suggestion
        toml.append({"modules": {"path": mod, "dependency_of": dependency_names}})
```

So perhaps a better name for `dependency_names` is `dependent_names`, but really... to avoid confusion between "dependancy" and "dependent", maybe we need other terms that are easier to tell apart.

stefanv replied:
This wording came from the SciPy PR; it's what their tool puts out. I thought it would be good to be compatible, but happy to talk about what this should look like.

mdhaber replied:
This comment can be closed; it's the code that needs to be changed.

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,127 @@
+
+
+def modules_dependent_on(modules: set[str] | list[str]) -> list[str]:
+    """Return the set of modules that depend on any of the given modules."""
```
```suggestion
    """Return the set of modules that are depended on by at least one of the given modules."""
```
If any of the changed `modules`, say module `j`, depend on another module, say module `i`, we want to run tests for module `i`.
So we'd say that an element of the returned set, say module `i`, is _depended on by_ one of the modules in the input list.
Another way we could put it:
```suggestion
    """Return the set of modules that is a dependency of at least one of the given modules."""
```

mdhaber reviewed (commented):
I think the logic is correct, but I think some of the variable names and text descriptions are mixed up. It's tricky because "dependency" and "dependent". How about "provider" and "consumer"?

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,127 @@
+
+    pkg_mods = _pkg_modules()
+    A = dependency_graph()
+    j = np.zeros_like(A[:, 0])  # boolean indices of dependent modules
+    for module in modules:
+        j |= A[:, pkg_mods.index(module)]
```
```suggestion
    A_j = np.zeros_like(A[:, 0])  # boolean indices of dependent modules
    for module in modules:
        A_j |= A[:, pkg_mods.index(module)]
```

stefanv committed:
Apply wording improvements by @mdhaber

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

stefanv commented:
In a bit of a :man_facepalming: moment, Matt just reminded me that, actually, we only, in theory, *need* to test modules that were modified: I mean, there test suite *should* be complete, and if they're not it's a bug. We don't need to test interdependencies, especially not within the library.

That said, we may try this as a safety layer for now:

1. Test module that was changed
2. Test all internal modules that rely on that module. **This should never fail.**
3. Test on main. **This should never fail.**

Eventually, we can get rid of (2).

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,130 @@
+        j = pkg_mods.index(module)
+        A_j |= A[:, j]
+
+    return sorted(set(np.array(pkg_mods, dtype=object)[j]))
```
```suggestion
    return sorted(set(np.array(pkg_mods, dtype=object)[A_j]))
```

stefanv committed:
Fix row reference by @mdhaber

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

stefanv committed:
Reformat

mdhaber commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,132 @@
+    A = dependency_graph()
+
+    toml = []
+    for j, mod in enumerate(mods):
+        dependency_names = mods_arr[A[:, j]].tolist()
```
I think the wording is correct. This is based on the fact that in the SciPy PR, `_lib` only "depends on" `_lib`, and everything else depends on `_lib`. That is the expected relationship.

```suggestion
    for i, mod in enumerate(mods):
        dependency_names = mods_arr[A[i, :]].tolist()
```

After changing this locally, I noticed that `skimage._shared` depends on several other modules, and I was surprised. I found at least one reason: in `_shared.testing`, there is `from .. import data, io`.

stefanv committed:
Split tests for modified module(s) and importers of those modules

stefanv committed:
Fix dependency list in text export

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

mdhaber commented on .github/workflows/test-autodepends.yaml:
```diff
@@ -0,0 +1,79 @@
+            echo "!!! Tests failed. This is **unexpected**, since the"
+            echo "!!! modified modules tests above passed."
+            echo "!!! You should find out what went wrong, and then update"
+            echo "!!! those modified module test suites to be more complete."
```
It's also possible that a feature or test in the unmodified module was incorrectly relying on a certain behavior of the function. For instance, a PR may *fix* a bug, but an outside test may have been relying on the faulty behavior, causing it to fail when the bug is fixed. Probably should acknowledge this possibility that it's not *always* the fault of the modified function.

mdhaber reviewed (approved):
Thanks for taking this from a half-baked idea to reality @stefanv! I think this is in pretty good shape. 

The time savings will depend on the module. For instance, changes to `_shared` will run tests for all modules because every other module uses *something* from `_shared`. These test runs will be a few seconds slower than they would be otherwise due to the time taken to compute the dependency graph. On the other hand, changes to modules like `graph` and `registration` will only require running tests for *those* modules, since nothing (see Details below) else imports from them, making test runs very fast. As an example of something in between, this PR makes a change to `morphology`. This causes tests in nine other modules, each of which relies on `morphology` somehow, to run. The slowest test run in this PR is less than 3 minutes, whereas the fastest run on macOS I've seen in other open PRs is at least 5 minutes.

<details>

It is possible that as-is, this misses some dependencies (e.g. imports within functions). In the worst case scenario, CI passes in a PR but fail after the PR merges and the full suite runs. 

This will be rare because four things have to happen: 
- there has to be a bug (or something else that would cause a test to fail), 
- the bug cannot be caught in review or by unit tests in the module $A$ being modfied, which will always run,
- this script does not detect that another module $B$ depends on $A$, and
- a test in $B$ must fail due to the change. 

In this case, an issue will be opened automatically (with gh-7752), and the PR can be amended or reverted, depending on the difficulty of the fix and severity of the disruption of test failures in main. The hole that allowed the dependency to slip through can be patched, making this sort of disruption even more rare, and beyond identifying a bug, this identifies a hole in unit tests for the feature being modified.

</details>

Besides the time savings, this PR makes another improvement by running tests for the modified module(s) *first*. Regardless of the module that is modified, this will dramatically reduce the time from pushing a commit to notification of most failures, which typically should be caught by tests in the module(s) being modified. In the case that tests in the modified module(s) *don't* fail but a test in another module *does*, the failure message provides a valuable reminder to look for holes in the unit test of the modified code. The project probably does not rely on unit tests of other projects for coverage; likewise, module $A$ should not rely on unit tests of module $B$ for coverage.

Thanks in advance to skimage maintainers for their feedback!

lagru commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,132 @@
+    mods = _pkg_modules()
+
+    n = len(mods)
+    A = np.zeros((n, n), dtype=bool)
```
A `dict[tuple[str, str], bool]` with the keys being module names might be a bit more "self-describing" and easier to debug?

But probably not worth the refactor. :)

stefanv replied:
Yes, but we use NumPy functionality in some places, so this would be harder.

lagru commented on tools/module_dependencies.py:
```diff
@@ -0,0 +1,132 @@
+    return A
+
+
+def dependency_toml():
```
I think I'm missing why this is named `toml` which is also used elsewhere? I don't see it serialized to a TOML file anywhere. Probably a left-over from refactoring?

stefanv replied:
This added a TOML file compatible with SciPy's checker, in case that became useful.

lucascolley replied:
Yeah, I think the idea was that something like https://github.com/scipy/scipy/blob/main/tach.toml could be substituted in in the future.

lagru commented on .github/workflows/test-autodepends.yaml:
```diff
@@ -0,0 +1,79 @@
+name: Run tests identified via dependency analysis
```
Nitpick, but I think having shorter names here helps the UX in quite a few places. E.g. the "Successful in 8m" isn't truncated in the PR dialog showing the CI checks, etc.

```suggestion
name: Test only changed
```

lagru commented on .spin/cmds.py:
```diff
@@ -81,3 +81,68 @@ def sdist(pyproject_build_args):
+            ['git', 'diff', 'main', '--stat', '--name-only'], output=False, echo=False
+        )
+        if p.returncode != 0:
+            raise (click.ClickException('Could not git-diff against main'))
```
Nitpick: What's with the enclosing parenthesis here? Weird that ruff didn't remove these since they seem unnecessary?

```suggestion
            raise click.ClickException('Could not git-diff against main')
```

lagru commented on .spin/cmds.py:
```diff
@@ -81,3 +81,68 @@ def sdist(pyproject_build_args):
+        pkg_mods = module_dependencies._pkg_modules()
+
+        p = spin.util.run(
+            ['git', 'diff', 'main', '--stat', '--name-only'], output=False, echo=False
```
Is it possible that this returns more files than necessary? Running this locally on the current version of the branch, it looks like this command also includes changes from commits on `main` that are not in `HEAD` too. E.g. `skimage/exposure/exposure.py` wasn't modified by this PR but is listed.

<details><summary>Details</summary>

```sh
git diff main --stat --name-only
```
```plain
.github/MAIN_FAIL_TEMPLATE.md
.github/workflows/emscripten.yaml
.github/workflows/test-autodepends.yaml
.github/workflows/test-docs.yaml
.github/workflows/test-linux.yaml
.github/workflows/test-macos.yaml
.github/workflows/test-nightlies-on-main.yaml
.github/workflows/test-windows.yaml
.github/workflows/wheels-recipe.yaml
.spin/cmds.py
CONTRIBUTING.rst
INSTALL.rst
TODO.txt
benchmarks/benchmark_restoration.py
doc/examples/applications/plot_coins_segmentation.py
doc/examples/applications/plot_thresholding_guide.py
doc/examples/features_detection/plot_remove_objects.py
doc/examples/filters/plot_deconvolution.py
doc/examples/segmentation/plot_mask_slic.py
doc/examples/segmentation/plot_metrics.py
doc/examples/segmentation/plot_regionprops.py
environment.yml
skimage/_shared/testing.py
skimage/_shared/tests/test_testing.py
skimage/exposure/exposure.py
skimage/feature/_basic_features.py
skimage/measure/_regionprops.py
skimage/measure/tests/test_regionprops.py
skimage/morphology/_flood_fill.py
skimage/morphology/misc.py
skimage/morphology/tests/test_misc.py
skimage/restoration/_cycle_spin.py
skimage/restoration/_rolling_ball.py
skimage/restoration/_rolling_ball_cy.pyx
skimage/restoration/tests/test_denoise.py
skimage/restoration/tests/test_rolling_ball.py
skimage/transform/_geometric.py
skimage/transform/_thin_plate_splines.py
skimage/transform/_warps.py
skimage/transform/tests/test_geometric.py
tools/generate_requirements.py
tools/module_dependencies.py
```


</details> 

Do we instead only want changes that are new in `HEAD` but not in `main`?

lagru reviewed (commented):
Okay, the idea looks simple enough. Except for a few nitpicks I have only one real comment left. 

Just so I understand correctly, this still needs to update other workflows so that only this one is run for PRs right?

lagru commented on .spin/cmds.py:
```diff
@@ -81,3 +81,68 @@ def sdist(pyproject_build_args):
+            to_test = to_test | importers_of_modified
+
+        pytest_args = kwargs.get('pytest_args', ())
+        if "--pyargs" in pytest_args:
```
The changes here now have somewhat major conflict with the recently introduced implementation for the `--doctest` option (#7786). Both override `--pyargs`.

lagru committed:
Merge branch 'main' into 'test-changed-module-dependencies'

stefanv commented:
FWIW easier alternative would be to just run test suite for module X that modified function resides in, but that requires that the test suite is "complete".

blink1073 commented:
> FWIW easier alternative would be to just run test suite for module X that modified function resides in, but that requires that the test suite is "complete".

IMO this is actually better because it introduces less complexity, and gives us a way to discover which test suites are "incomplete".   I'd be more hesitant if we didn't have the auto-issue-creation steps on failure.

stefanv commented:
I'm happy to go that direction too, @blink1073. This was a really fun and interesting experiment to do with Matt, but the ultimate goal remains simply to vastly reduce CI runtime.

mdhaber commented:
That sounds fine. The original was a balance between speed and avoiding red CI workflows in main. But if a project is confident enough in their intra-module tests to skip the stepping stone, great!

stefanv commented:
This work is being continued in https://github.com/scikit-image/scikit-image/pull/8103
